File: ios-native/Sources/ALCore/Support/RandomSource.swift
// The injected, seedable randomness — the Swift shape of every `Math.random()`
// in `src/levels.ts`, plus the ordered-dedup helper that keeps a seeded run
// reproducible.
//
// D9: randomness is INJECTED AND SEEDABLE, and there is deliberately NO shared
// global instance. Two reasons, both learned from the React original:
//
//   1. A round builder that cannot be replayed cannot be tested, and these
//      builders carry the whole difficulty curve.
//   2. In React the session is built once, in a lazy `useState` initialiser. In
//      SwiftUI an expression in a `@State` default is evaluated EVERY time the
//      view struct is initialised even though only the first value is kept — so
//      building a session there silently builds and discards one on every parent
//      re-render. With a system RNG that is invisible; with a shared seeded one
//      it destroys reproducibility. Hence: never call a `build…Session` from a
//      view's `init` or a `@State` default (build it in `.task {}` or in an
//      `@Observable` model created once), and never reach for a global source.
//
// The convenience defaults below are `= .system()`, which Swift evaluates per
// call — a FRESH source each time, never a shared one.

/// The primitive every source provides: 64 raw bits.
///
/// Trivial on purpose — it exists so `RandomSource` can box any base without
/// caring whether it is the system generator or a deterministic one.
public protocol ALRandomGenerator {
    mutating func nextBits() -> UInt64
}

/// Deterministic, portable, no Foundation. SplitMix64 — a handful of mixing
/// steps per draw, and stable across OS versions (`arc4random` is not, and
/// `SystemRandomNumberGenerator` is not reproducible at all).
public struct SeededGenerator: ALRandomGenerator {
    private var state: UInt64

    public init(seed: UInt64) {
        self.state = seed
    }

    public mutating func nextBits() -> UInt64 {
        state = state &+ 0x9E37_79B9_7F4A_7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58_476D_1CE4_E5B9
        z = (z ^ (z >> 27)) &* 0x94D0_49BB_1331_11EB
        return z ^ (z >> 31)
    }
}

/// Production randomness. Not reproducible, by definition.
public struct SystemGenerator: ALRandomGenerator {
    private var base = SystemRandomNumberGenerator()

    public init() {}

    public mutating func nextBits() -> UInt64 { base.next() }
}

/**
 * One run's randomness, threaded explicitly through every builder that consumes
 * it (always as the LAST parameter, so the call sites read like the TS did).
 */
// NB: ARCHITECTURE.md §5 sketches `RandomSource` as a `protocol` with a single
// `mutating func next(upperBound: Int) -> Int`. It is implemented here as the
// final class that data-core.md §4.1 specifies, for the reasons given there:
// `spellIntruders`, `buildGridRound` and `buildTwinRound` all call helpers and
// closures while holding partial arrays, and threading `inout` through that is
// noise at every level; and "the session's RNG" is naturally a reference. The
// `next(upperBound:)` spelling from ARCHITECTURE.md exists below as a method, so
// a call site written against either document compiles.
//
// Deliberately NOT `Sendable`: it is mutable reference state touched only from
// the UI thread. Fine in language mode v5 (D1); when ALCore flips to v6 this
// becomes `@MainActor`, a two-line change — provided nobody made it a global
// singleton in the meantime.
public final class RandomSource {
    private var draw: () -> UInt64

    public init<G: ALRandomGenerator>(_ base: G) {
        var g = base
        self.draw = { g.nextBits() }
    }

    /// A fresh, non-reproducible source. The default for every builder.
    public static func system() -> RandomSource { RandomSource(SystemGenerator()) }

    /// A reproducible source. The whole point of the injection.
    public static func seeded(_ seed: UInt64) -> RandomSource { RandomSource(SeededGenerator(seed: seed)) }

    /// `0 ..< n`, unbiased (Lemire's multiply-shift with rejection).
    /// Returns 0 when `n <= 0`. The TS analogue is `(Math.random() * n) | 0`.
    public func int(below n: Int) -> Int {
        guard n > 0 else { return 0 }
        let bound = UInt64(n)
        var product = draw().multipliedFullWidth(by: bound)
        if product.low < bound {
            // (2^64 mod bound), computed without 128-bit arithmetic.
            let threshold = (0 &- bound) % bound
            while product.low < threshold {
                product = draw().multipliedFullWidth(by: bound)
            }
        }
        return Int(product.high)
    }

    /// ARCHITECTURE.md §5's spelling of `int(below:)`. Same function.
    public func next(upperBound n: Int) -> Int { int(below: n) }

    /// The TS analogue is `Math.random() < 0.5`.
    public func bool() -> Bool { int(below: 2) == 0 }

    /**
     * Exact port of `shuffle<T>` from `src/levels.ts`:
     *
     *     const a = arr.slice();
     *     for (let i = a.length - 1; i > 0; i--) {
     *       const j = (Math.random() * (i + 1)) | 0;
     *       [a[i], a[j]] = [a[j], a[i]];
     *     }
     *
     * Written out rather than delegated to `Array.shuffled(using:)` on purpose:
     * descending Fisher–Yates with one draw per step means the NUMBER OF DRAWS
     * PER SHUFFLE matches the TS, which keeps a seeded Swift run structurally
     * comparable to a seeded JS run. The stdlib uses the same algorithm but
     * draws differently.
     */
    public func shuffled<T>(_ arr: [T]) -> [T] {
        var a = arr
        guard a.count > 1 else { return a }
        var i = a.count - 1
        while i > 0 {
            let j = int(below: i + 1)
            a.swapAt(i, j)
            i -= 1
        }
        return a
    }

    /// `nil` on an empty array. Analogue of `a[(Math.random() * a.length) | 0]`,
    /// whose `undefined` the TS then `??`-defaults.
    public func element<T>(of a: [T]) -> T? {
        guard !a.isEmpty else { return nil }
        return a[int(below: a.count)]
    }
}

/**
 * Insertion-ordered dedupe — the Swift shape of `Array.from(new Set(xs))`.
 */
// NB: this helper is not a convenience, it is a CORRECTNESS requirement.
// JS `Set` iterates in insertion order; Swift `Set` is unordered AND randomised
// per process. Three sites depend on the order (data-core.md §4.3):
//   - `Content.syllableBank` — its order IS the distractor distribution
//     (`pickDistractorSyllable` picks by index) and its first element is the
//     fallback (today "CHA");
//   - `spellIntruders`, which walks the answer's distinct letters in
//     first-appearance order;
//   - `buildFirstLetterSession`'s letter catalog.
// Missing one makes the result nondeterministic UNDER A FIXED SEED, which reads
// as a flaky test rather than a port bug. A plain `Set` is fine for pure
// membership tests (`need.has`, `seen.has`) — those never iterate.
public func orderedUnique<T: Hashable>(_ xs: [T]) -> [T] {
    var seen = Set<T>()
    var out: [T] = []
    out.reserveCapacity(xs.count)
    for x in xs where seen.insert(x).inserted {
        out.append(x)
    }
    return out
}

File: apps/game-android/core/src/main/kotlin/fr/dappit/attrapelettres/core/support/RandomSource.kt
package fr.dappit.attrapelettres.core.support

import kotlin.random.Random

/* -------------------------------------------------------------------------- */
/* The injected, seedable randomness — the Kotlin shape of every                */
/* `Math.random()` in `src/levels.ts`, plus the ordered-dedup helper that keeps */
/* a seeded run reproducible.                                                   */
/*                                                                              */
/* Randomness is INJECTED AND SEEDABLE, and there is deliberately NO shared     */
/* global instance (ARCHITECTURE.md §5: "injected, never global"). A round      */
/* builder that cannot be replayed cannot be tested, and these builders carry   */
/* the whole difficulty curve. The default arguments on the builders are        */
/* `= SystemRandomSource()`, which Kotlin evaluates per call — a FRESH source   */
/* each time, never a shared one.                                               */
/* -------------------------------------------------------------------------- */

/**
 * One run's randomness, threaded explicitly through every builder that
 * consumes it (always as the LAST parameter, so the call sites read like the
 * TS did). `next` is the single requirement; everything the builders actually
 * call — [bool], [shuffled], [elementOf] — is derived from it below, so any
 * implementation is automatically consistent across all four.
 */
interface RandomSource {
    /** `0 until upperBound`, uniform. Returns 0 when `upperBound <= 0` — the TS analogue `(Math.random() * n) | 0` yields 0 there (`NaN | 0`), and the call sites rely on never trapping. */
    fun next(upperBound: Int): Int
}

/**
 * Deterministic, portable, dependency-free. SplitMix64 — a handful of mixing
 * steps per draw, stable across JVMs and OS versions (`java.util.Random`'s
 * algorithm is specified but its seeding conventions invite accidents, and
 * `kotlin.random.Random(seed)`'s algorithm is not documented as frozen). The
 * same generator the iOS port uses, for the same reason: a seeded run must
 * mean the same thing next year.
 */
class SeededGenerator(seed: Long) : RandomSource {
    private var state: ULong = seed.toULong()

    /** The raw 64-bit draw. */
    private fun nextBits(): ULong {
        state += 0x9E37_79B9_7F4A_7C15uL
        var z = state
        z = (z xor (z shr 30)) * 0xBF58_476D_1CE4_E5B9uL
        z = (z xor (z shr 27)) * 0x94D0_49BB_1331_11EBuL
        return z xor (z shr 31)
    }

    override fun next(upperBound: Int): Int {
        if (upperBound <= 0) return 0
        val bound = upperBound.toULong()
        // arc4random_uniform-style rejection: `0uL - bound` wraps to
        // 2^64 - bound, whose remainder is 2^64 mod bound — the size of the
        // sliver at the bottom of the range that would make `% bound` uneven.
        // Unbiased with plain 64-bit arithmetic; no 128-bit multiply, which the
        // JVM only grew portably after our Android floor.
        val threshold = (0uL - bound) % bound
        var bits = nextBits()
        while (bits < threshold) {
            bits = nextBits()
        }
        return (bits % bound).toInt()
    }
}

/** Production randomness. Not reproducible, by definition. */
class SystemRandomSource : RandomSource {
    override fun next(upperBound: Int): Int =
        if (upperBound <= 0) 0 else Random.nextInt(upperBound)
}

/** The TS analogue is `Math.random() < 0.5`. */
fun RandomSource.bool(): Boolean = next(2) == 0

/**
 * Exact port of `shuffle<T>` from `src/levels.ts`:
 *
 *     const a = arr.slice();
 *     for (let i = a.length - 1; i > 0; i--) {
 *       const j = (Math.random() * (i + 1)) | 0;
 *       [a[i], a[j]] = [a[j], a[i]];
 *     }
 *
 * Written out rather than delegated to `List.shuffled(Random)` on purpose:
 * descending Fisher–Yates with one draw per step means the NUMBER OF DRAWS PER
 * SHUFFLE matches the TS, which keeps a seeded Kotlin run structurally
 * comparable to a seeded JS run. The stdlib uses the same algorithm but draws
 * through a different interface, so a seed would not survive the substitution.
 */
fun <T> RandomSource.shuffled(list: List<T>): List<T> {
    if (list.size < 2) return list.toList()
    val a = list.toMutableList()
    var i = a.size - 1
    while (i > 0) {
        val j = next(i + 1)
        val tmp = a[i]
        a[i] = a[j]
        a[j] = tmp
        i -= 1
    }
    return a
}

/**
 * `null` on an empty list. Analogue of `a[(Math.random() * a.length) | 0]`,
 * whose `undefined` the TS then `??`-defaults.
 */
fun <T> RandomSource.elementOf(list: List<T>): T? =
    if (list.isEmpty()) null else list[next(list.size)]

/**
 * Insertion-ordered dedupe — the shape of `Array.from(new Set(xs))`.
 *
 * Kotlin's `distinct()` already preserves first-seen order (it walks through a
 * `LinkedHashSet`), so unlike the Swift port this is not dodging a randomised
 * `Set` — but the named helper stays, for two reasons. The three ports share a
 * vocabulary, so the sites that are ORDER-SENSITIVE announce themselves the
 * same way in every language: `SYLLABLE_BANK` (its order IS the
 * distractor distribution and its first element is the fallback, today "CHA"),
 * `spellIntruders` (walks the answer's distinct letters in first-appearance
 * order) and the first-letter catalog. And the property is pinned by our own
 * test rather than by stdlib documentation — a future "optimisation" to a
 * plain `Set` would fail loudly here instead of surfacing as a flaky seeded
 * run.
 */
fun <T> orderedUnique(xs: List<T>): List<T> = xs.distinct()
